api/v1alpha1/amaltheasession_types.go

```diff
@@ -0,0 +1,265 @@
+	// The format of the configuration that is expected depends on the storage type.
+	// NOTE: define all values in a single key of the Kubernetes secret.
+	// rclone: any valid rclone configuration for a single remote, see the output of `rclone config providers` for validation and format.
+	SecretRef string `json:"secretRef,omitempty"`
+}
+
+type Culling struct {
+	// +kubebuilder:default:=0
+	// +kubebuilder:validation:Minimum:=0
+	// The maximum allowed age for a session, regardless of whether it
+	// is active or not. When the threshold is reached the session is hibernated.
+	// A value of zero indicates that Amalthea will not automatically hibernate
+	// the session based on its age.
+	MaxAgeSeconds int `json:"maxAgeSeconds,omitempty"`
+	// +kubebuilder:default:=0
+	// +kubebuilder:validation:Minimum:=0
+	// How long should a server be idle for before it is hibernated. A value of
+	// zero indicates that Amalthea will not automatically hibernate inactive sessions.
+	IdleSeconds int `json:"idleSeconds,omitempty"`
+	// +kubebuilder:default:=0
+	// +kubebuilder:validation:Minimum:=0
+	// How long can a server be in starting state before it gets hibernated. A
+	// value of zero indicates that the server will not be automatically hibernated
+	// by Amalthea because it took to long to start.
+	StartingSeconds int `json:"startingSeconds,omitempty"`
+	// +kubebuilder:default:=0
+	// +kubebuilder:validation:Minimum:=0
+	// How long can a server be in failed state before it gets hibernated. A
+	// value of zero indicates that the server will not be automatically
+	// hibernated by Amalthea if it is failing.
+	FailedSeconds int `json:"failedSeconds,omitempty"`
+	// +kubebuilder:default:=0
+	// +kubebuilder:validation:Minimum:=0
+	// Number of seconds where a server can be in hibernated state before
+	// it gets completely deleted. A value of zero indicates that hibernated servers
+	// will not be automatically be deleted by Amalthea after a period of time.
+	HibernatedSeconds int `json:"hibernatedSeconds,omitempty"`
+}
+
+// +kubebuilder:validation:Enum={token,oauth2proxy}
+type AuthenticationType string
+
+const Token AuthenticationType = "token"
+const Oidc AuthenticationType = "oauth2proxy"
+
+type Authentication struct {
+	Enabled bool               `json:"enabled,omitempty"`
+	Type    AuthenticationType `json:"type,omitempty"`
+	// Kubernetes secret that contains the authentication configuration
+	// For `token` generate a hard to guess string / password-like string.
+	// this value can be used as Authorization header or as a cookie with the name `amaltheaSessionToken` to
+	// access the session.
+	// For `oauth2proxy` please see https://oauth2-proxy.github.io/oauth2-proxy/configuration/overview#config-file.
+	SecretRef SecretRef `json:"secretRef,omitempty"`
+}
+
+// A reference to a Kubernetes secret and a specific field in the secret to be used in a session
+type SecretRef struct {
+	Name string `json:"name"`
+	Key  string `json:"key"`
+}
```

Comment:
Shouldn't we use [LocalObjectReference](https://github.com/kubernetes/api/blob/e09016fffd8ee1edc6575b8ea7c314e9c398120f/core/v1/types.go#L6647) rather than Name like
[SecretEnvSource](https://github.com/kubernetes/api/blob/e09016fffd8ee1edc6575b8ea7c314e9c398120f/core/v1/types.go#L2371) does ?
Granted LocalObjectReference currently has a single field called Name but it might be more explicit in what is expected.

Reply:
So I embedded `LocalObjectReference` and this is what the relevant CRD sections looks like:

```
                  secretRef:
                    description: Kubernetes secret that contains the authentication
                      configuration For `token` generate a hard to guess string /
                      password-like string. this value can be used as Authorization
                      header or as a cookie with the name `amaltheaSessionToken` to
                      access the session. For `oauth2proxy` please see https://oauth2-proxy.github.io/oauth2-proxy/configuration/overview#config-file.
                    properties:
                      key:
                        description: The secret field or key name to be used.
                        type: string
                      name:
                        description: 'Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
                          TODO: Add other useful fields. apiVersion, kind, uid?'
                        type: string
                    type: object
                    x-kubernetes-map-type: atomic

```

Because the docs on the `Name` field in `LocalObjectReference` are a bit weird.

I think it is better to stick with what we have above.

Reply:
Fair enough, let's keep it.